### PR TITLE
Defer search for dev cert until build bind time

### LINKED ISSUE #45801: Kestrel config behaves differently when set using env vars vs. in-code
I have a Web API application that I want to configure it to listen on an HTTPS port. The app is containerized and orchestrated using Docker compose.

I can configure it by setting the following env variable in `docker-compose`, which works as expected.

```yaml
services:
  webapi:
    environment:
      - ASPNETCORE_URLS=https://+:443
```

However, when I use the following approach as an alternative to setting the environment variable, it fails with the following error message.

```csharp
builder.WebHost.ConfigureKestrel(options =>
{
    options.Listen(
    address: IPAddress.Any,
        port: 443,
        configure: listenOptions =>
        {
            listenOptions.Protocols = HttpProtocols.Http1AndHttp2;
            listenOptions.UseHttps();
        });
});
```

> Unable to configure HTTPS endpoint. No server certificate was specified, and the default developer certificate could not be found or is out of date.

When I provide the filename and password to the SSL certificate file as the following, it works as expected.

```csharp
builder.WebHost.ConfigureKestrel(options =>
{
    options.Listen(
    address: IPAddress.Any,
        port: 443,
        configure: listenOptions =>
        {
            listenOptions.Protocols = HttpProtocols.Http1AndHttp2;
            listenOptions.UseHttps("/root/.aspnet/https/cert.pfx", "Password");
        });
});
```

I assume setting the environment variable `ASPNETCORE_URLS` or calling `ConfigureKestrel` are equivalent, though, I am not sure what I am missing in the second approach. 

### Update 
I was previously getting the following error, which was related to mapping the port `443` to an incorrect port of the container.

> Detected a TLS handshake to an endpoint that does not have TLS enabled.

### PR DESCRIPTION
**Pull request by amcasey**
# Defer search for dev cert until build bind time

- [x] You've read the [Contributor Guide](https://github.com/dotnet/aspnetcore/blob/main/CONTRIBUTING.md) and [Code of Conduct](https://github.com/dotnet/aspnetcore/blob/main/CODE-OF-CONDUCT.md).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

## Description

...so that it can occur after IConfiguration is read. (In particular, so that it occurs during AddressBinder.BindAsync which happens strictly after KestrelConfigurationLoader.Load.) This is important in Docker scenarios since the Docker tools use IConfiguration to tell us where the dev cert directory is mounted.# with '#' will be ignored, and an empty message aborts the commit.

Was #46296
Fixes #45801

### CODE REVIEW

**Comment by amcasey** [comment]
The behavior is finalized (subject to approval), but we can't merge this until #48055 is approved and #48054 is merged and adopted by dotnet-monitor or we'll break them again.

File: src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs [hunk 1]
```diff
@@ -188,17 +191,29 @@ public static ListenOptions UseHttps(this ListenOptions listenOptions, Action<Ht
...
 
+        // NB: This lambda will only be invoked if either HTTP/1.* or HTTP/2 is being used
         listenOptions.Use(next =>
         {
-            // Set the list of protocols from listen options
-            httpsOptions.HttpProtocols = listenOptions.Protocols;
-            var middleware = new HttpsConnectionMiddleware(next, httpsOptions, loggerFactory, metrics);
+            // Evaluate the HttpsConnectionAdapterOptions, now that the configuration, if any, has been loaded
+            var httpsOptions = lazyHttpsOptions.Value;
+            var loggerFactory = listenOptions.KestrelServerOptions.ApplicationServices.GetRequiredService<ILoggerFactory>();
```
Comment (mitchdenny) [inline src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:214]: Will these service lookups here have a performance implication?

Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:214]: It could be moved outside of the lambda so it's not happening per connection.

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:214]: +1. They were outside the lambda before.

Comment (amcasey) [reply src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:214]: For my own edification, this is a squeeze-out-every-drop type optimization, right?  We're saving a lookup in a (small) dictionary and a null check?

Comment (Tratcher) [reply src/Servers/Kestrel/Core/src/ListenOptionsHttpsExtensions.cs:214]: It's a no-need-to-do-duplicate-work cleanup. We don't expect it to have a huge per impact, but there's no reason to do that extra work here, it's no cleaner.

**Comment by amcasey** [comment]
Will try eagerly loading configuration instead.

**Pull request closed without merging**